pkg/tempo/servicegraphprocessor/processor.go

```diff
@@ -0,0 +1,251 @@
+	logger := log.With(util.Logger, "component", "tempo service graphs")
+
+	if cfg.Wait == 0 {
+		cfg.Wait = DefaultWait
+	}
+	if cfg.MaxItems == 0 {
+		cfg.MaxItems = DefaultMaxItems
+	}
+
+	// TODO(mapno): Add support for an external cache (e.g. memcached)
+	p := &processor{
+		nextConsumer: nextConsumer,
+		reg:          prometheus.DefaultRegisterer,
+		// Cleanup period is hardcoded to twice the waiting time for simplicity
+		// Most likely not ideal in every scenario
+		store:    cache.New(cfg.Wait, cfg.Wait*2),
+		maxItems: cfg.MaxItems,
+		closed:   atomic.Bool{},
+		logger:   logger,
+	}
+
+	return p, nil
+}
+
+func (p *processor) Start(context.Context, component.Host) error {
+	return p.registerMetrics()
+}
+
+func (p *processor) registerMetrics() error {
+	p.serviceGraphRequestTotal = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "tempo_service_graph_request_total",
+		Help: "Total count of requests between two nodes",
+	}, []string{"client", "server"})
+	p.serviceGraphRequestFailedTotal = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "tempo_service_graph_request_failed_total",
+		Help: "Total count of failed requests between two nodes",
+	}, []string{"client", "server"})
+	p.serviceGraphRequestHistogram = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name:    "tempo_service_graph_request_seconds",
+		Help:    "Time for a request between two nodes",
+		Buckets: prometheus.ExponentialBuckets(0.01, 2, 12),
+	}, []string{"client", "server"})
+	p.serviceGraphUnpairedSpansTotal = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "tempo_service_graph_unpaired_spans_total",
+		Help: "Total count of requests between two nodes",
+	}, []string{"client", "server"})
+
+	cs := []prometheus.Collector{
+		p.serviceGraphRequestTotal,
+		p.serviceGraphRequestFailedTotal,
+		p.serviceGraphRequestHistogram,
+		p.serviceGraphUnpairedSpansTotal,
+	}
+
+	for _, c := range cs {
+		if err := p.reg.Register(c); err != nil {
+			return err
+		}
+	}
+
+	// Collect unpaired spans when evicting items from the store during
+	// periodic cleanup
+	p.store.OnEvicted(func(s string, i interface{}) {
+		e := i.(edge)
+		if !e.complete() {
+			p.serviceGraphUnpairedSpansTotal.WithLabelValues(e.clientService, e.serverService).Inc()
+		}
+	})
+
+	return nil
+}
+
+func (p *processor) Shutdown(context.Context) error {
+	p.closed.Store(true)
+	p.unregisterMetrics()
+	return nil
+}
+
+func (p *processor) unregisterMetrics() {
+	cs := []prometheus.Collector{
+		p.serviceGraphRequestTotal,
+		p.serviceGraphRequestFailedTotal,
+		p.serviceGraphRequestHistogram,
+		p.serviceGraphUnpairedSpansTotal,
+	}
+
+	for _, c := range cs {
+		p.reg.Unregister(c)
+	}
+}
+
+func (p *processor) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{}
+}
+
+func (p *processor) ConsumeTraces(ctx context.Context, td pdata.Traces) error {
+	level.Info(p.logger).Log("msg", "consuming traces")
+	if p.closed.Load() {
```

Comment:
do we need this? the otel shutdown shuts down receivers first and then processors.

Reply:
Addressed in 7122551971a9044c815835d8f8b4c8d65b0385c2